File: src/main/scala/com/github/phisgr/gatling/grpc/request/CallDefinition.scala
package com.github.phisgr.gatling.grpc.request

import com.github.phisgr.gatling.generic.check.ResponseExtract
import com.github.phisgr.gatling.grpc
import com.github.phisgr.gatling.grpc.HeaderPair
import com.github.phisgr.gatling.grpc.check.CheckWithSelfType
import com.github.phisgr.gatling.grpc.protocol.GrpcProtocol
import io.gatling.commons.validation.{Success, Validation}
import io.gatling.core.action.builder.ActionBuilder
import io.gatling.core.check.CheckBuilder.{Find, MultipleFind}
import io.gatling.core.session.{Expression, Session}
import io.grpc.{CallOptions, Metadata, MethodDescriptor}

/**
 * Trait to reuse code for specifying checks and call options.
 *
 * @tparam Self Curiously Recurring Template Pattern
 * @tparam Req  request
 * @tparam Res  response
 */
trait CallDefinition[Self, Req, Res] extends ActionBuilder {
  // putting Self as a type member makes type inference dumber

  /**
   * Wraps around `Res` for checking.
   * `GrpcResponse[Res]` (which includes status and trailers) for client streams and unary calls;
   * simply `Res` for server and bidi-streams.
   */
  type Wrap[_]
  type Check[R] <: CheckWithSelfType[Wrap[R], Check[R]]

  private[gatling] def checks: List[Check[Res]]

  def check(checks: Check[Res]*): Self

  def checkIf(condition: Expression[Boolean])(thenChecks: Check[Res]*): Self =
    check(thenChecks.map(_.checkIf(condition)): _*)

  def checkIf(condition: (Wrap[Res], Session) => Validation[Boolean])(thenChecks: Check[Res]*): Self =
    check(thenChecks.map(_.checkIf(condition)): _*)

  // In fact they can be added to checks using .check
  // but the type Res cannot be inferred there
  def extract[X](
    f: Res => Validation[Option[X]])(
    ts: (Find[ResponseExtract, Res, X] => Check[Res])*
  ): Self = {
    val e = grpc.Predef.extract(f)
    check(mapToList(ts)(_.apply(e)): _*)
  }

  def extractIf[X](
    condition: Expression[Boolean])(
    f: Res => Validation[Option[X]])(
    ts: (Find[ResponseExtract, Res, X] => Check[Res])*
  ): Self =
    extract(f)(ts.map(_.andThen(_.checkIf(condition))): _*)

  def extractIf[X](
    condition: (Wrap[Res], Session) => Validation[Boolean])(
    f: Res => Validation[Option[X]])(
    ts: (Find[ResponseExtract, Res, X] => Check[Res])*
  ): Self =
    extract(f)(ts.map(_.andThen(_.checkIf(condition))): _*)

  def extractMultiple[X](
    f: Res => Validation[Option[Seq[X]]])(
    ts: (MultipleFind[ResponseExtract, Res, X] => Check[Res])*
  ): Self = {
    val e = grpc.Predef.extractMultiple[Res, X](f)
    check(mapToList(ts)(_.apply(e)): _*)
  }

  def extractMultipleIf[X](
    condition: Expression[Boolean])(
    f: Res => Validation[Option[Seq[X]]])(
    ts: (MultipleFind[ResponseExtract, Res, X] => Check[Res])*
  ): Self =
    extractMultiple(f)(ts.map(_.andThen(_.checkIf(condition))): _*)

  def extractMultipleIf[X](
    condition: (Wrap[Res], Session) => Validation[Boolean])(
    f: Res => Validation[Option[Seq[X]]])(
    ts: (MultipleFind[ResponseExtract, Res, X] => Check[Res])*
  ): Self =
    extractMultiple(f)(ts.map(_.andThen(_.checkIf(condition))): _*)

  private def mapToList[T, U](s: Seq[T])(f: T => U): List[U] = s.map(f).toList


  private[gatling] def callAttributes: CallAttributes

  private[gatling] def withCallAttributes(callAttributes: CallAttributes): Self

  def callOptions(callOptions: Expression[CallOptions]): Self =
    withCallAttributes(callAttributes.copy(callOptions = callOptions))

  /**
   * Sets call options of the call.
   *
   * @param callOptions is a call-by-name param, it will be run every time the call is executed.
   * @return a new GrpcCallActionBuilder
   */
  def callOptions(callOptions: => CallOptions): Self =
    this.callOptions { _: Session => Success(callOptions) }

  def header[T](key: Metadata.Key[T])(value: Expression[T]): Self =
    withCallAttributes(callAttributes.copy(
      reversedHeaders = HeaderPair(key, value) :: callAttributes.reversedHeaders
    ))

  def target(protocol: GrpcProtocol): Self =
    this.withCallAttributes(callAttributes.copy(protocolOverride = Some(protocol)))

  private[gatling] val method: MethodDescriptor[Req, Res]
}

File: kt/src/main/kotlin/com/github/phisgr/gatling/kt/grpc/request/CallDefinition.kt
package com.github.phisgr.gatling.kt.grpc.request

import com.github.phisgr.gatling.grpc.check.CheckWithSelfType
import com.github.phisgr.gatling.kt.grpc.GrpcProtocol
import com.github.phisgr.gatling.kt.grpc.internal.*
import io.gatling.commons.validation.Validation
import io.gatling.javaapi.core.ActionBuilder
import io.gatling.javaapi.core.CheckBuilder
import io.gatling.javaapi.core.Session
import io.gatling.javaapi.core.internal.Expressions
import io.grpc.CallOptions
import io.grpc.Metadata
import scala.Function2
import com.github.phisgr.gatling.grpc.request.CallDefinition as CallDefinitionS
import io.gatling.core.session.Session as SessionS

@Suppress("UNCHECKED_CAST")
abstract class CallDefinition<
    Self : CallDefinition<Self, Req, Res, WrappedRes, Wrapped, Check>,
    Req,
    Res,
    WrappedRes,
    Wrapped : CallDefinitionS<Wrapped, Req, Res>,
    Check : CheckWithSelfType<WrappedRes, Check>,
    > : ActionBuilder {
    abstract override fun asScala(): Wrapped

    abstract fun wrap(wrapped: Wrapped): Self

    @JvmName("_KtHeader")
    inline fun <reified T> header(key: Metadata.Key<T>, el: String): Self = header(key, el, T::class.java)

    inline fun <T> header(key: Metadata.Key<T>, crossinline value: (Session) -> T): Self = wrap(
        asScala().header(key, toScalaExpression { session: SessionS ->
            value(Session(session))
        })
    )

    fun <T> header(key: Metadata.Key<T>, el: String, clazz: Class<T>): Self = wrap(
        asScala().header(key, Expressions.toExpression(el, clazz))
    )

    inline fun callOptions(crossinline callOptions: (Session) -> CallOptions): Self = wrap(
        asScala().callOptions(toScalaExpression { session: SessionS ->
            callOptions(Session(session))
        })
    )


    /**
     * To define the extractions outside this method,
     * see [com.github.phisgr.gatling.kt.grpc.extract] and
     * [com.github.phisgr.gatling.kt.grpc.extractMultiple].
     */
    @JvmName("_KotlinCheck")
    fun check(vararg checks: From<Res>.() -> CheckBuilder): Self = wrap(
        asScala().check(
            checks.map { buildCheck(it.invoke(dummyFrom)) }.toSeq()
        )
    )

    fun check(vararg checks: CheckBuilder): Self = wrap(
        asScala().check(
            checks.map { buildCheck(it) }.toSeq()
        )
    )

    fun target(protocol: GrpcProtocol) = wrap(asScala().target(protocol.protocol()))

    fun checkIf(condition: String) = ConditionWithoutRes(
        this as Self,
        Expressions.toBooleanExpression(condition)
    )

//    fun checkIf(condition: Function<Session, Boolean>) = ConditionWithoutRes(
//        this as Self,
//        Expressions.javaBooleanFunctionToExpression(condition)
//    )
//
//    fun checkIf(condition: BiFunction<WrappedRes, Session, Boolean>): ConditionWithRes<Self, Req, Res, WrappedRes, Wrapped, Check> =
//        checkIf { res, session -> condition.apply(res, session) }

    //    @JvmName("_KotlinCheckIf")
    inline fun checkIf(crossinline condition: (Session) -> Boolean) = ConditionWithoutRes(
        this as Self
    ) { session -> boolValidation { condition(Session(session)) } }

    //    @JvmName("_KotlinCheckIf")
    inline fun checkIf(crossinline condition: (WrappedRes, Session) -> Boolean) = ConditionWithRes(
        this as Self
    ) { res, session ->
        boolValidation { condition(res as WrappedRes, Session(session)) }
    }

    protected abstract fun buildCheck(check: CheckBuilder): Check

    class ConditionWithoutRes<
        Self : CallDefinition<Self, Req, Res, WrappedRes, Wrapped, Check>,
        Req,
        Res,
        WrappedRes,
        Wrapped : CallDefinitionS<Wrapped, Req, Res>,
        Check : CheckWithSelfType<WrappedRes, Check>,
        >(
        private val builder: Self,
        private val condition: scala.Function1<SessionS, Validation<UnknownTypeParam>>,
    ) {
        @JvmName("_KotlinThen")
        fun then(vararg checks: From<Res>.() -> CheckBuilder): Self =
            then(*checks.map { it.invoke(dummyFrom) }.toTypedArray())

        fun then(vararg checks: CheckBuilder): Self = builder.wrap(
            builder.asScala().checkIf(
                condition,
                checks.map { builder.buildCheck(it) }.toSeq()
            )
        )
    }

    class ConditionWithRes<
        Self : CallDefinition<Self, Req, Res, WrappedRes, Wrapped, Check>,
        Req,
        Res,
        WrappedRes,
        Wrapped : CallDefinitionS<Wrapped, Req, Res>,
        Check : CheckWithSelfType<WrappedRes, Check>,
        >(
        private val builder: Self,
        private val condition: Function2<UnknownTypeParam, SessionS, Validation<UnknownTypeParam>>,
    ) {
        @JvmName("_KotlinThen")
        fun then(vararg checks: From<Res>.() -> CheckBuilder): Self =
            then(*checks.map { it.invoke(dummyFrom) }.toTypedArray())

        fun then(vararg checks: CheckBuilder): Self = builder.wrap(
            builder.asScala().checkIf(
                condition,
                checks.map { builder.buildCheck(it) }.toSeq()
            )
        )
    }
}
